### PR TITLE
http-task: allow automatic redirects for POST requests

### PR DESCRIPTION
**Pull request by d0s056f**
post requests in http task should follow page redirects (instead of returning 301) if task parameter `followRedirects` is set to `true` or not provided in the request

https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3

### CODE REVIEW

**Comment by d0s056f** [comment]
@ibodrov does this require an IT?

**Comment by benbroadaway** [comment]
I think this should an option, but not enabled by default since it's _not_ the default behavior of the HTTP spec.

**Comment by ibodrov** [comment]
> @ibodrov does this require an IT?

Not really, but there are existing unit tests for the task. Might be a good idea to add a unit test for this feature too.

**Comment by d0s056f** [comment]
> I think this should an option, but not enabled by default since it's _not_ the default behavior of the HTTP spec.

@benbroadaway so another flag in the task parameters? something like `allowPostRedirect`?

**Comment by ibodrov** [comment]
There's already `followRedirects` option, which is enabled by default.
While I personally think it should've been `false`, we don't have much choice there.
Let POST use the same option and leave it `true` by default, I guess?

**Comment by benbroadaway** [comment]
Yeah we're kind of in a pickle with the existing `followRedirects` option's behavior versus what should be the default behavior for redirecting POST calls and expanding the total number of parameters. It should be okay to tie it to the `followRedirects` option and make it clear in the documentation that it will follow redirects for each type of call.

**Comment by d0s056f** [comment]
okay. just adding the unit test then

**Comment by d0s056f** [comment]
https://github.com/walmartlabs/concord-website/pull/32

**Comment by ibodrov** [comment]
Getting
```
04:11:50 [INFO ] [ERROR] Tests run: 38, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 8.569 s <<< FAILURE! - in com.walmartlabs.concord.plugins.http.HttpTaskTest
04:11:50 [INFO ] [ERROR] testExecuteGetRequestWithFollowRedirect(com.walmartlabs.concord.plugins.http.HttpTaskTest)  Time elapsed: 0.051 s  <<< ERROR!
04:11:50 [INFO ] org.apache.http.conn.HttpHostConnectException: Connect to localhost:8001 [localhost/127.0.0.1] failed: Connection refused (Connection refused)
04:11:50 [INFO ] Caused by: java.net.ConnectException: Connection refused (Connection refused)
04:11:50 [INFO ] 
04:11:50 [INFO ] [ERROR] testExecutePostRequestWithFollowRedirect(com.walmartlabs.concord.plugins.http.HttpTaskTest)  Time elapsed: 0.037 s  <<< ERROR!
04:11:50 [INFO ] org.apache.http.conn.HttpHostConnectException: Connect to localhost:8001 [localhost/127.0.0.1] failed: Connection refused (Connection refused)
04:11:50 [INFO ] Caused by: java.net.ConnectException: Connection refused (Connection refused)
```

**Comment by d0s056f** [comment]
@ibodrov it's failing because of this stub - https://github.com/walmartlabs/concord/pull/122/files#diff-3900108df33fe2d401f4378aab0de0e4R289
worked on my local as there's a local concord server running on 8001. Should I replace this with a generic endpoint like google .com?

**Comment by ibodrov** [comment]
I think it should be a wiremock endpoint.

**Comment by d0s056f** [comment]
does this work? sorry still learning :)

**Comment by benbroadaway** [comment]
@d0s056f The wiremock-y way to generate the URL is like
```java
        rule.stubFor(post(urlEqualTo("/followRedirectsPost"))
                .willReturn(permanentRedirect(rule.url("/string")))
        );
```

File: plugins/tasks/http/src/main/java/com/walmartlabs/concord/plugins/http/SimpleHttpClient.java [hunk 1]
```diff
@@ -332,6 +333,7 @@ private static CloseableHttpClient createClient(Configuration cfg) throws Except
         return HttpClientBuilder.create()
                 .setConnectionManager(buildConnectionManager())
                 .setDefaultRequestConfig(c.build())
+                .setRedirectStrategy(new LaxRedirectStrategy())
```
Comment (benbroadaway) [inline plugins/tasks/http/src/main/java/com/walmartlabs/concord/plugins/http/SimpleHttpClient.java:336]: This should be enabled based on the `followRedirects` option. Something like:
```java
RedirectStrategy redirectStrategy;
if (cfg.isFollowRedirects()) {
    redirectStrategy = new LaxRedirectStrategy();
} else {
    redirectStrategy = DefaultRedirectStrategy.INSTANCE;
}

return HttpClientBuilder.create()
        .setConnectionManager(buildConnectionManager())
        .setDefaultRequestConfig(c.build())
        .setRedirectStrategy(redirectStrategy)
        .build();
```

**Review by benbroadaway — CHANGES_REQUESTED** [review]

File: plugins/tasks/http/src/main/java/com/walmartlabs/concord/plugins/http/SimpleHttpClient.java (hunk 1 above)
Comment (d0s056f) [reply plugins/tasks/http/src/main/java/com/walmartlabs/concord/plugins/http/SimpleHttpClient.java:336]: @benbroadaway I have tested this. it works only when followRedirects is true, and doesn't when false. you can check the test cases too

Comment (benbroadaway) [reply plugins/tasks/http/src/main/java/com/walmartlabs/concord/plugins/http/SimpleHttpClient.java:336]: Aaah so it's already tied to `redirectsEnabled` in the apache `RequestConfig`. Got it.

**Review by benbroadaway — APPROVED** [review]

**Comment by ibodrov** [comment]
@concord-ci test it

**Pull request merged**